### PR TITLE
chore: fixes for leanprover/lean4#2790

### PR DESCRIPTION
**Pull request by kim-em**
leanprover/lean4#2790 changes how some internal exceptions (in particular, max recursion depth errors) are handled.

Some are now surfacing here!

The problem in `Archive` seem to be new and interesting: some calls to `simp` or `norm_num` are failing with max recursion depth errors.

Previously, these errors were being caught, and ... *something* ... was then trying a different path and succeeding. Now the error is stopping progress.

---

I'm hoping a filter/notation expert might be able to have a look at this to find a better solution than my workaround here.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
I've created two zulip threads:
* https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/request.20for.20help.20with.20filter.20notation
* https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/request.20for.20help.20with.20failure.20in.20Archive
to ask for assistance with the two problems here.

File: Mathlib/Topology/Basic.lean [hunk 1]
```diff
@@ -848,7 +848,7 @@ scoped[Topology] notation "𝓝" => nhds
 scoped[Topology] notation "𝓝[" s "] " x:100 => nhdsWithin x s
 
 /-- Notation for the filter of punctured neighborhoods of a point. -/
-scoped[Topology] notation "𝓝[≠] " x:100 => nhdsWithin x {x}ᶜ
+scoped[Topology] notation "𝓝[≠] " x:100 => nhdsWithin x (@singleton _ (Set _) instSingletonSet x)ᶜ
```
Comment (eric-wieser) [inline Mathlib/Topology/Basic.lean:851]: Does the following work?
```suggestion
scoped[Topology] notation "𝓝[≠] " x:100 => nhdsWithin x ({x}ᶜ : Set _)
```

Comment (PatrickMassot) [reply Mathlib/Topology/Basic.lean:851]: Eric, you can see from the git history that your proposition was my first "fix", but it didn't work.

Comment (kmill) [reply Mathlib/Topology/Basic.lean:851]: I'm seeing `nhdsWithin x ({x} : Set _)ᶜ` in the git history -- Eric's suggesting `nhdsWithin x ({x}ᶜ : Set _)`

Comment (kmill) [reply Mathlib/Topology/Basic.lean:851]: I've tried a few things and I can't get it to work like that.

I did find this workaround though:
```
set_option quotPrecheck false in
/-- Notation for the filter of punctured neighborhoods of a point. -/
scoped[Topology] notation "𝓝[≠] " x:100 => letI z := x; nhdsWithin z ({z}ᶜ)
```
I'm not sure what's going on or why this helps at this point.

File: lake-manifest.json [hunk 2]
```diff
@@ -40,5 +40,13 @@
     "opts": {},
     "name": "proofwidgets",
     "inputRev?": "v0.0.21",
+    "inherited": false}},
+  {"git":
+   {"url": "https://github.com/leanprover/std4",
```
Comment (jcommelin) [inline lake-manifest.json:45]: Std4 also has an entry at the top of the file. But the `inputRev?` field is different. Is this quasi-duplication intentional?

Comment (semorrison) [reply lake-manifest.json:45]: No, an accident, now resolved.

**Comment by semorrison** [comment]
Ugh, this is going to be impossible to rebuild, without rebasing the (already merged!) leanprover/lean4#2790 onto a more recent nightly.

File: Archive/Imo/Imo1960Q1.lean [hunk 3]
```diff
@@ -95,7 +95,12 @@ theorem right_direction {n : ℕ} : ProblemPredicate n → SolutionPredicate n :
   have := searchUpTo_start
   iterate 82
     replace :=
-      searchUpTo_step this (by norm_num1; rfl) (by norm_num1; rfl) (by norm_num1; rfl) (by norm_num)
+      searchUpTo_step this (by norm_num1; rfl) (by norm_num1; rfl) (by norm_num1; rfl)
+        (by -- This used to be just `norm_num`, but after leanprover/lean4#2790,
+            -- that triggers a max recursion depth exception. As a workaround, we
+            -- manually rewrite with `Nat.digits_of_two_le_of_pos` first.
+            rw [Nat.digits_of_two_le_of_pos (by norm_num) (by norm_num)]
+            norm_num)
```
Comment (eric-wieser) [inline Archive/Imo/Imo1960Q1.lean:103]: Presumably this is a place where increasing the depth doesn't help?

**Pull request closed without merging**